File: leetcode/medium/Subsets.java
import java.util.ArrayList;
import java.util.List;

/**
 * Solution to generate all possible subsets (the power set) of a set of unique integers.
 */
class Solution {
    
    /**
     * Generates all subsets using an iterative bitmask approach.
     * This avoids the overhead of recursion and allows pre-allocation of the result list.
     * * Time Complexity: O(n * 2^n) - We iterate through 2^n combinations.
     * Space Complexity: O(n * 2^n) - To store the result.
     */
    public List<List<Integer>> subsets(int[] nums) {
        int n = nums.length;
        // The number of subsets is 2^n
        int numberOfSubsets = 1 << n; 
        List<List<Integer>> result = new ArrayList<>(numberOfSubsets);

        for (int i = 0; i < numberOfSubsets; i++) {
            List<Integer> subset = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                // Check if the j-th bit of i is set
                if ((i & (1 << j)) != 0) {
                    subset.add(nums[j]);
                }
            }
            result.add(subset);
        }
        return result;
    }
}

File: leetcode/medium/Subsets.py
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        """
        Generates the Power Set (all possible subsets) of a given integer array.
        
        Complexity:
        - Time: O(n * 2^n) - There are 2^n subsets, and we iterate up to n times for each.
        - Space: O(n * 2^n) - To store all subsets in the result list.
        """
        n = len(nums)
        # The total number of subsets is 2^n
        subsets_count = 1 << n 
        result = []

        # Iterate through every number from 0 to 2^n - 1
        for i in range(subsets_count):
            current_subset = []
            # Check each bit position j
            for j in range(n):
                # If the j-th bit of 'i' is set, include nums[j]
                if (i >> j) & 1:
                    current_subset.append(nums[j])
            result.append(current_subset)

        return result